### ISSUE TITLE
pcp daemons issuing warnings to tty on which they were updated

### ISSUE THREAD

**Issue by filbo** [performancecopilot/pcp#577]
Hi,

This report is about old, back-rev versions of everything, and as such may be completely superseded and invalid.  I am not in a position to upgrade pcp on the system in question.  So this is intended more or less as a query: were you ever aware of this issue; did you fix it already; if not, maybe you would like to.

It may also be entirely a distro-specific packaging error.  However, there is at least one change worth making in the daemon startup code (see below).

First, this is on LinuxMint 17.3 'rosa', which is based on Ubuntu 14.04 'trusty', based on Debian 8 'jessie/sid'.

It's about an installation & startup behavior which I believe is wrong.  You will probably also feel that some of *my* behaviors are wrong; be that as it may.

This is my desktop system and as you can see, I haven't wholesale upgraded it in a while.  So it's running pcp as delivered by the distro chain: in this case, the version is given as '3.8.12ubuntu1' so I'm pretty sure it's out of Ubuntu.

Now, not only do I not upgrade, I am also extremely averse to rebooting.  At the moment the system has been up for 210 days.  This means that I have a lot of ancient daemons running, including pcp.  I also have a number of extremely persistent xterms (well, mate-terminals).

I generally run updates (`apt-get update; apt-get dist-upgrade`, more or less) in a particular xterm which I consider my 'admin xterm'.

In that xterm, I am now getting this message every night around 2am:

    pmlogextract: Warning: timezone mismatch for input archives
    archive: /var/log/pcp/pmmgr/r5omino/merged-archive-20181120.100307 timezone: PDT+7 [will be used]
    archive: /var/log/pcp/pmmgr/r5omino/archive-20181120.100308 timezone: PST+8 [will be ignored]

This started after the timezone change, i.e. November 5th.  The nature of the message seems obvious and is not my concern.

The misbehavior is that the messages are being delivered to the tty on which an update was run.  These are daemons which are normally started during system startup and will not be attached to any tty.  When they're restarted during an update, the combination of package & init scripts should arrange for that same environment, i.e. not attached to any particular tty.  Otherwise there can be some message leakage, and possibly some sort of more significant security consequence (though no particular one occurs to me off the top of my head).

The system has been up since 2018-04-25.  `apt` logs show that on 2018-05-04, all of the pcp packages were updated; this is what caused the daemons to be restarted from my admin xterm.  (The apt logs actually show the pcp packages being updated FROM version '3.8.12ubuntu1' TO '3.8.12ubuntu1' -- notice these are the same strings.  In fact all of the files /var/cache/apt/archives/\*pcp\*.deb are dated 2014-01-29, so I have no idea why it went through this gyration.  But it did.)

I reason out the fault as follows:

- each daemon has been invoked with a '-l logfile' flag; `lsof` shows that the respective files are open as fd 2 (stderr), but fd 1 points to my admin tty (bad).  (fd 0 is /dev/null, which is reasonable)

- exception: despite a '-l' flag, `pmmgr` does *NOT* have a log file open; both stdout and stderr point to my admin tty (worse)

- although pmmgr(1) doesn't mention it, pmlogger(1) says:

    If the log file cannot be created or is not writable, output is written to standard error instead.

- so `pmmgr` has failed to open its log file despite the correct-looking '-l' flag (but `lsof` shows its cwd to be /var/log/pcp/pmmgr, which directory exists and is owned by user 'pcp', which the daemon is running under; so I can't explain how / why it failed to create it)

- it is probably `pmmgr` which is invoking `pmlogextract` to generate the warning

- I verified that `pmlogextract` issues the warning to stderr (which further implicates `pmmgr`):

    $ pmlogextract /var/log/pcp/pmmgr/\`hostname -s\`/*.0 foo.0 > stdout-file 2> stderr-file
    -- now 'stdout-file' is empty while 'stderr-file' has the timezone warning

- FAULT: the update scripts in the .deb package (probably pcp.postinst) should not pass any vestige of the tty on which the update was run.  This means its invocation of *all* of the daemons' init scripts should be wrapped in something along the lines of `setsid ...init-cmd... < /dev/null > /dev/null 2>&1`

- probably NOT FAULT: the init scripts themselves should not do any such thing.  One might be intentionally invoking them in some sort of semi-interactive mode for testing purposes.  During system startup, they're run by startup scripts which themselves have no tty and are probably directing all of the stdio fds to /dev/null.

- NOT FAULT: the daemons themselves shouldn't know anything about this stuff; they should be correctly invoked by their various scripts

- HOWEVER: it is perhaps unwise to allow the original passed-in stdout to dangle in the wind.  The common startup code of all daemons in the project should take control of stdout, in case any invoked subprocess tries to write to it.  Either open /dev/null as stdout, or dup stderr over it (which would be the '-l' logfile if successfully opened, else the original passed-in stderr)

NOTE: it is NORMAL Debian => Ubuntu => Mint behavior that daemons are restarted after their packages are updated.  The mere fact of updating a package which contains an in-use daemon does *NOT* force or require a reboot.  It is expected that daemon-bearing packages deal with this (and to a nigh-remarkable degree, they do).

**Comment by fche** [comment]
I cannot explain your observations with pmmgr.  It has apprx. always supported both -l and non-l modes, with no related changes since pcp 3.8.12.  In -l mode, both stdout and stderr are redirected.  With a systemd platform, relaying diagnostic output to stdout and/or stderr is preferable anyway.

**Comment by filbo** [comment]
This LinuxMint is not yet in the `systemd` era.  (Current LinuxMint is; I am backrev.)

So again, I am reporting:

- one fault that you probably won't do anything about (incorrect preconfig of stdout / stderr in the old sysv startup scripts -- therefore passing the tty of an upgrade process into a long-lived system daemon)

- one fault whereby started daemons may write error messages to the passed-in stdout

I haven't looked at the code, but I now suspect the nature of the 2nd fault is this:

```
if (lflag) {
   stderr := open(lflag_filename)
   stdout := stderr
}
```

should be:

```
if (lflag) {
   stderr := open(lflag_filename)
}
stdout := stderr
```

i.e.: **always** forcibly dup stderr onto stdout, whether or not the -l flag was specified or the given filename was successfully opened.  Don't carry around the original stdout when anything coming out of any part of the system (including subprograms your project may not have written) is necessarily some sort of error message or unexpected condition.

**Comment by filbo** [comment]
I mean only when starting a daemon qua daemon (i.e.: canceled by any sort of 'debug' or 'run in foreground' flag).

I'm not sure if I have it right.  I've just left the pcp set of daemons running, just like the update left them; do not care to mess around with restarting them.  You have the domain knowledge:

Observe what happens if you start the entire pcp subsystem from a terminal, with a '-l' flag specifying an openable file (e.g. /var/log/pcp/test.log).  Look at the fds of the resulting tree of daemons.  I think in this case you should see no instances of the terminal on fds 0 / 1 / 2, only test.log or /dev/null.

Then try with a '-l' flag pointing to a non-openable file (/does/not/exist).  Look at the fds of the resulting tree of daemons.  Some are /dev/null, some are /dev/pts/terminal-you-were-on, in a non-sensible pattern.

Once started that way, what happens if you exit that terminal, leaving /dev/pts/2 (or whatever) in a blocked state.  Do the daemons exit immediately?  Only the ones that had the terminal open?  Do they not exit immediately, but some time later when they try to write to that fd?  Even later, after trying to write to it for a while, until some buffer in the pty driver fills up?  Instead of exiting, do they eventually wedge, blocked on that filled pty buffer?

I don't think you want any of those outcomes.

**Comment by fche** [comment]
> Observe what happens if you start the entire pcp subsystem from a terminal ...

Your observations are correct, but relate to environments where a user takes control of launching the suite of software, but then fails to ensure that the software can send diagnostics.  It's not unlike worrying about a user ^Z suspending a process of her own, or exiting with a SIGHUP when the user logged out.  Yes, it can happen, but a user kind of did it to herself.

Services started properly with an init system of some sort are given log files & file handles that are not going to be stuck this way.  Services started by hand, and ran in a normal way (without suspending, without exiting the terminal, et.c) will work fine too.

**Comment by filbo** [comment]
I did not manually launch, I performed a normal system update (on this system, effectively, `apt-get update; apt-get dist-upgrade`).  If I've followed the files correctly, this causes /var/lib/dpkg/info/pcp.prerm to run:
```
invoke-rc.d pmproxy stop
invoke-rc.d pmwebd stop
invoke-rc.d pmie stop
invoke-rc.d pmlogger stop
invoke-rc.d pmcd stop
```
-- then, later, /var/lib/dpkg/info/pcp.postinst to run:
```
invoke-rc.d pmcd start
invoke-rc.d pmlogger start
```
(this asymmetry is present in the respective files)

They are being invoked by the local flavor of init system, yet end up with dangling references to the tty on which the update was run.
